MapleServer2/Types/Stats.cs

```diff
@@ -0,0 +1,545 @@
+                StatId.MinAtk,
+                new Stat(0)
+            },
+            {
+                StatId.MaxAtk,
+                new Stat(0)
+            },
+            {
+                StatId.MinDmg,
+                new Stat(0)
+            },
+            {
+                StatId.MaxDmg,
+                new Stat(0)
+            },
+            {
+                StatId.Pierce,
+                new Stat(0)
+            },
+            {
+                StatId.MountSpeed,
+                new Stat(100)
+            },
+            {
+                StatId.BonusAtk,
+                new Stat(0)
+            },
+            {
+                StatId.PetBonusAtk,
+                new Stat(0)
+            }
+        };
+    }
+
+    public Stat this[StatId statIndex]
+    {
+        get => Data[statIndex];
+
+        private set => Data[statIndex] = value;
+    }
+
+    public void InitializePools(int hp, int spirit, int stamina)
+    {
+        Stat hpStat = this[StatId.Hp];
+        Stat spiritStat = this[StatId.Spirit];
+        Stat staStat = this[StatId.Stamina];
+        Data[StatId.Hp] = new(hpStat.Bonus, hpStat.Base, hp);
+        Data[StatId.Spirit] = new(spiritStat.Bonus, spiritStat.Base, spirit);
+        Data[StatId.Stamina] = new(staStat.Bonus, staStat.Base, stamina);
+    }
+
+    public void Allocate(StatId statId)
+    {
+        int gainAmount = 1;
+        switch (statId)
+        {
+            case StatId.Hp:
+                gainAmount = 10;
+                break;
+            case StatId.CritRate:
+                gainAmount = 3;
+                break;
+            default:
+                break;
+        }
+        Data[statId].IncreaseBonus(gainAmount);
+    }
+
+    public void ResetAllocations(StatDistribution statDist)
+    {
+        foreach (KeyValuePair<byte, int> entry in statDist.AllocatedStats)
+        {
+            StatId statId = (StatId) entry.Key;
+            int gainAmount = 1;
+            switch (statId)
+            {
+                case StatId.Hp:
+                    gainAmount = 10;
+                    break;
+                case StatId.CritRate:
+                    gainAmount = 3;
+                    break;
+                default:
+                    break;
+            }
+            Data[statId].DecreaseBonus(entry.Value * gainAmount);
+        }
+        statDist.ResetPoints();
+    }
+}
+
+/* Max = Base + stat allocations + bonuses.
+ * Min = Base stat amount.
+ * Current = Final value (e.g. capped Damage, current Hp, active CritRate, ...)
+ *
+ * Change PlayerStat.Current for temporary changes.
+ */
+public class Stat
+{
+    public long BonusLong;
+    public long BaseLong;
+    public long TotalLong;
+    public int Bonus { get => (int) BonusLong; protected set { BonusLong = value; } }
+    public int Base { get => (int) BaseLong; protected set { BaseLong = value; } }
+    public int Total { get => (int) TotalLong; protected set { TotalLong = value; } }
```

Comment:
Why this is needed when everything is long? Same for the methods below

Reply:
Doing this allows us to use a stat in 1 type for dictionaries and as a int or as a long without type casting. We can't use class inheritance because it won't serialize.